File: AdventOfCode2022/11/11b.cpp
#include <bits/stdc++.h> //only use this for Comp-Prog

using namespace std;
#define loop(i,n) for(ll i = 0; i < n; i++)
typedef unsigned long long ll;
ll doOp(ll old, const char op, ll num) {
	if(op=='+') return old + num;
	if(op=='*') return old * num;
	if(op=='^') return old * old;
}
void reset(std::stringstream& stream)
{
    const static std::stringstream initial;
    
    stream.str(std::string());
    stream.clear();
    stream.copyfmt(initial);
}
int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	ll n = 8;
	vector<char> ops = {'*', '+', '+', '+', '+', '*', '^', '+'};
	vector<ll> nums = {2,   3,   6,   5,   8,   5,   2,   4};
	// ll n = 4;
	// vector<char> ops = {'*', '+', '^', '+'};
	// vector<ll> nums = {19, 6, 2, 3};
	vector<ll> div(n),t(n),f(n);
	vector<queue<ll>> v(n);
	string tmp,line;
	ll bruh = 0;
	loop(i,n) {
		bruh = i*100;
		getline(cin,line); //Monkey #
		getline(cin,line); //start 
		stringstream ss(line);
		ss>>tmp>>tmp; //starting items:
		while(ss>>tmp) { //takes in numbers
			v[i].push(stoll(tmp));
			// cout<<stoll(tmp)<<endl;
		}
		getline(cin,line); //operation
		reset(ss); 
		ss<<line;
		ss>>tmp>>tmp>>tmp>>tmp; //Operation: new = old
		ss>>tmp>>tmp; //operator, number (hard coded)

		getline(cin,line); //test
		reset(ss);
		ss<<line;
		ss>>tmp>>tmp>>tmp; //Test: divisible by
		ss>>tmp;
		div[i]=stoll(tmp);

		getline(cin,line); //if true
		reset(ss);
		ss<<line;
		ss>>tmp>>tmp>>tmp>>tmp>>tmp; //if true: throw to monkey
		ss>>tmp;
		t[i]=stoll(tmp);
		getline(cin,line); //if false
		reset(ss);
		ss<<line;
		ss>>tmp>>tmp>>tmp>>tmp>>tmp; // if false: throw to monkey
		ss>>tmp;
		f[i]=stoll(tmp);
		getline(cin,line);
	}
	vector<ll> touch(n,0);
	loop(z,10000) {
		loop(i,n) {
			while(!v[i].empty()) {
				touch[i]++;
				ll worry = v[i].front();
				v[i].pop();
				worry = doOp(worry, ops[i], nums[i]);
				ll mod = 9699690;
				if(worry%div[i]==0) {
					v[t[i]].push(worry%mod);
				} else {
					v[f[i]].push(worry%mod);
				}
			}
		}
	}
	for(ll i = 0; i < n; i++) {
		cout<<"Monkey "<<i<<": ";
		while(v[i].empty()==0) {
			cout<<v[i].front()<< " ";
			v[i].pop();
		}
		cout<<endl;
	}
	std::sort(touch.rbegin(), touch.rend());
	cout<<touch[0]*touch[1]<<endl;
	return 0;	
}

File: AdventOfCode2022/11/11b.py

from queue import Queue
#functions to make input simpler, taken from https:#codeforces.com/blog/entry/83441 
def inp():
   # convert input to int and return 
   return(int(input()))
def inlt():
   # split input (demarkers whitespace), convert to int, and return list of ints
   return(list(map(int,input().split())))
def insr():
   # remove leading and trailing whitespace and return input (string) 
   return(input().strip())
def invr():
   # convert input to int (delimiter is whitespace) and return map object
   return(map(int,input().split()))


def doOp(old, op, num):
	if(op=='+'):
		return old + num
	elif(op=='*'):
		return old * num
	elif(op=='^'):
		return old * old

def main():
	n = 8
	ops = ['*', '+', '+', '+', '+', '*', '^', '+']
	nums = [2,   3,   6,   5,   8,   5,   2,   4]

	# n = 4
	# ops = ['*', '+', '^', '+']
	# nums = [19, 6, 2, 3]


	div = [0]*n
	t = [0]*n
	f = [0]*n
	file = open("11.txt")
	v = [Queue(maxsize = 40) for i in range(n)]
	for i in range(n):
		line = file.readline() #Monkey #
		line = file.readline().strip().split() #start 
		for j in range(2,len(line)):
			if line[j][-1]==",":
				line[j] = line[j].replace(',','')
			v[i].put(int(line[j]))

		line = file.readline() #operation
		 #Operation: new = old
		 #operator, number (hard coded)

		line = file.readline() #test
		f1,f2,f3,mod = line.strip().split() #Test: divisible by
		div[i]=int(mod)

		line = file.readline().strip().split() #if true
		t[i]=int(line[-1])


		line = file.readline().strip().split() #if false
		f[i]=int(line[-1])
		line = file.readline() #empty
	
	touch = [0]*n
	for z in range(10000):
		if z%10==0:
			print(z)
		for i in range(n):
			while v[i].empty()==False:
				touch[i] += 1
				worry = v[i].get()
				
				
				worry = doOp(worry, ops[i], nums[i])
				if worry%div[i]==0:
					v[t[i]].put(worry)
				else:
					v[f[i]].put(worry)
	touch.sort()
	print(touch[-1]*touch[-2])
main()